sdk/azcore/internal/shared/expiring_resource.go

```diff
@@ -59,8 +64,9 @@ func (er *ExpiringResource) GetResource(state interface{}) (interface{}, error)
 			// Getting here means that this thread/goroutine will wait for the updated resource
 		} else if er.expiration.Add(-window).Before(now) {
 			// The resource is valid but is expiring within the time window
-			if !er.acquiring {
-				// If another thread/goroutine is not acquiring/renewing the resource, this thread/goroutine will do it
+			if !er.acquiring && er.lastAcquired.Add(backoff).Before(now) {
```

Comment:
The test fails because the callback returns `time.Now()` as the value's expiration time. When clock resolution is low or the runtime is fast (apparently one of these is the case on Windows), the second `GetResource()`  gets the same value from `time.Now()`, finds the current time equal to the expiration time, and doesn't refresh because the last attempt was less than 30 seconds prior. I think it's a test bug--if the test intends its first resource value to be expired on arrival, it should return a past expiration time for that value.

As for whether `GetResource` should refresh in this case anyway, I think it shouldn't. The STS counts all requests equally. The intent of the 30 second wait is to avoid hitting a rate limit with an unnecessary request. In this case the resource hasn't expired, so a refresh is unnecessary.